### ISSUE TITLE
[Inductor] [CPU] `nn.LayerNorm` outputs numerical error compared with eager

### ISSUE THREAD

**Issue by shaoyuyoung** [pytorch/pytorch#143182]
### 🐛 Describe the bug

related to #142839.

To be honest, I am not sure whether it's my negligence because It seems that many `Norm` operators output numerical error in `Inductor mode`.  

This error is only obvious on `CPU`.

Subsequently, I would check other `Norm` operators.

Plz feel free to correct me if I am wrong. :)

```python
import torch
import torch.nn as nn


class Model(nn.Module):

    def __init__(self):
        super().__init__()
        self.ln = nn.LayerNorm([10000, 1000])  # LayerNorm for 2D input

    def forward(self, x):
        x = self.ln(x)
        return x


model = Model().eval()

x = torch.randn(1, 3, 10000, 1000)  # As `H` and `W` increase, the error might be amplified

inputs = [x]

c_model = torch.compile(model)

output = model(*inputs)

c_output = c_model(*inputs)

print(torch.allclose(output, c_output, 1e-5, 1e-5))  # loose check in fp32
print(torch.max(torch.abs(output - c_output)))

fp_64_ref = c_model(x.double())
print("Eager divergence", torch.max(torch.abs(output - fp_64_ref)))
print("Compile divergence divergence", torch.max(torch.abs(c_output - fp_64_ref)))

```

### Error logs

cpu
```
False
tensor(0.0007, grad_fn=<MaxBackward1>)
Eager divergence tensor(6.1528e-07, dtype=torch.float64, grad_fn=<MaxBackward1>)
Compile divergence divergence tensor(0.0007, dtype=torch.float64, grad_fn=<MaxBackward1>)
```

### Versions

torch version: 2.6.0.dev20241205+cu124
OS: Ubuntu 20.04.6 LTS (x86_64)
CPU: Intel(R) Xeon(R) Gold 6248 CPU @ 2.50GHz
GPU: V100

<details>
<summary>click for detailed env</summary>

```
PyTorch version: 2.6.0.dev20241205+cu124
Is debug build: False
CUDA used to build PyTorch: 12.4
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.6 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.2) 9.4.0
Clang version: 16.0.1
CMake version: version 3.26.0
Libc version: glibc-2.31

Python version: 3.12.7 | packaged by Anaconda, Inc. | (main, Oct  4 2024, 13:27:36) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.4.0-200-generic-x86_64-with-glibc2.31
Is CUDA available: True
CUDA runtime version: 12.6.68
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: Tesla V100-SXM2-32GB
GPU 1: Tesla V100-SXM2-32GB
GPU 2: Tesla V100-SXM2-32GB
GPU 3: Tesla V100-SXM2-32GB

Nvidia driver version: 560.35.03
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.9.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv.so.9.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn.so.9.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_engines_precompiled.so.9.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_engines_runtime_compiled.so.9.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_graph.so.9.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_heuristic.so.9.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops.so.9.6.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Byte Order:                         Little Endian
Address sizes:                      40 bits physical, 48 bits virtual
CPU(s):                             20
On-line CPU(s) list:                0-19
Thread(s) per core:                 1
Core(s) per socket:                 20
Socket(s):                          1
NUMA node(s):                       1
Vendor ID:                          GenuineIntel
CPU family:                         6
Model:                              85
Model name:                         Intel(R) Xeon(R) Gold 6248 CPU @ 2.50GHz
Stepping:                           7
CPU MHz:                            2500.000
BogoMIPS:                           5000.00
Hypervisor vendor:                  KVM
Virtualization type:                full
L1d cache:                          640 KiB
L1i cache:                          640 KiB
L2 cache:                           80 MiB
L3 cache:                           16 MiB
NUMA node0 CPU(s):                  0-19
Vulnerability Gather data sampling: Unknown: Dependent on hypervisor status
Vulnerability Itlb multihit:        KVM: Vulnerable
Vulnerability L1tf:                 Mitigation; PTE Inversion
Vulnerability Mds:                  Vulnerable: Clear CPU buffers attempted, no microcode; SMT Host state unknown
Vulnerability Meltdown:             Mitigation; PTI
Vulnerability Mmio stale data:      Vulnerable: Clear CPU buffers attempted, no microcode; SMT Host state unknown
Vulnerability Retbleed:             Mitigation; IBRS
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; IBRS; IBPB conditional; STIBP disabled; RSB filling; PBRSB-eIBRS Not affected; BHI SW loop, KVM SW loop
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Vulnerable: Clear CPU buffers attempted, no microcode; SMT Host state unknown
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon rep_good nopl xtopology cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch topoext cpuid_fault invpcid_single pti ssbd ibrs ibpb fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm mpx avx512f avx512dq rdseed adx smap clflushopt clwb avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves arat umip pku ospke avx512_vnni

Versions of relevant libraries:
[pip3] numpy==1.26.4
[pip3] nvidia-cublas-cu12==12.4.5.8
[pip3] nvidia-cuda-cupti-cu12==12.4.127
[pip3] nvidia-cuda-nvrtc-cu12==12.4.127
[pip3] nvidia-cuda-runtime-cu12==12.4.127
[pip3] nvidia-cudnn-cu12==9.1.0.70
[pip3] nvidia-cufft-cu12==11.2.1.3
[pip3] nvidia-curand-cu12==10.3.5.147
[pip3] nvidia-cusolver-cu12==11.6.1.9
[pip3] nvidia-cusparse-cu12==12.3.1.170
[pip3] nvidia-cusparselt-cu12==0.6.2
[pip3] nvidia-nccl-cu12==2.21.5
[pip3] nvidia-nvjitlink-cu12==12.4.127
[pip3] nvidia-nvtx-cu12==12.4.127
[pip3] optree==0.13.1
[pip3] pytorch-triton==3.2.0+git35c6c7c6
[pip3] torch==2.6.0.dev20241205+cu124
[pip3] torchaudio==2.5.0.dev20241205+cu124
[pip3] torchvision==0.20.0.dev20241205+cu124
[pip3] triton==3.0.0
[conda] numpy                     1.26.4                   pypi_0    pypi
[conda] nvidia-cublas-cu12        12.4.5.8                 pypi_0    pypi
[conda] nvidia-cuda-cupti-cu12    12.4.127                 pypi_0    pypi
[conda] nvidia-cuda-nvrtc-cu12    12.4.127                 pypi_0    pypi
[conda] nvidia-cuda-runtime-cu12  12.4.127                 pypi_0    pypi
[conda] nvidia-cudnn-cu12         9.1.0.70                 pypi_0    pypi
[conda] nvidia-cufft-cu12         11.2.1.3                 pypi_0    pypi
[conda] nvidia-curand-cu12        10.3.5.147               pypi_0    pypi
[conda] nvidia-cusolver-cu12      11.6.1.9                 pypi_0    pypi
[conda] nvidia-cusparse-cu12      12.3.1.170               pypi_0    pypi
[conda] nvidia-cusparselt-cu12    0.6.2                    pypi_0    pypi
[conda] nvidia-nccl-cu12          2.21.5                   pypi_0    pypi
[conda] nvidia-nvjitlink-cu12     12.4.127                 pypi_0    pypi
[conda] nvidia-nvtx-cu12          12.4.127                 pypi_0    pypi
[conda] optree                    0.13.1                   pypi_0    pypi
[conda] pytorch-triton            3.2.0+git35c6c7c6          pypi_0    pypi
[conda] torch                     2.6.0.dev20241205+cu124          pypi_0    pypi
[conda] torchaudio                2.5.0.dev20241205+cu124          pypi_0    pypi
[conda] torchvision               0.20.0.dev20241205+cu124          pypi_0    pypi
[conda] triton                    3.0.0                    pypi_0    pypi
```

</details>

cc @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @aakhundov

**Comment by IvanKobzarev** [comment]
@shaoyuyoung 
If to check `test/inductor` folder, most of thests are using abs and relative tolerance ~ `1e-3`
With 1e-3 your model passes, so I do not put this as high-pri.

**Comment by eellison** [comment]
Hi @shaoyuyoung, please compare the results to an fp64 reference input and only submit an issue if compile diverges more from fp64 than eager. 

```
fp_64_ref = c_model(x.double())
print("Eager divergence", torch.max(torch.abs(output - fp_64_ref)))
print("Compile divergence divergence", torch.max(torch.abs(c_output - fp_64_ref)))

Eager divergence tensor(0.0001, device='cuda:0', dtype=torch.float64, grad_fn=<MaxBackward1>)
Compile divergence divergence tensor(4.9160e-07, device='cuda:0', dtype=torch.float64,
       grad_fn=<MaxBackward1>)
``` 
As you can see, the results are closer, so we are more accurately computing this.


cc @lazos - can you submit a public version of the fuzzer thing you wrote ? this would be great to reference and clear up theses sorts of issues - fallback_random as well.

**Comment by mlazos** [comment]
> Hi @shaoyuyoung, please compare the results to an fp64 reference input and only submit an issue if compile diverges more from fp64 than eager.
> 
> ```
> fp_64_ref = c_model(x.double())
> print("Eager divergence", torch.max(torch.abs(output - fp_64_ref)))
> print("Compile divergence divergence", torch.max(torch.abs(c_output - fp_64_ref)))
> 
> Eager divergence tensor(0.0001, device='cuda:0', dtype=torch.float64, grad_fn=<MaxBackward1>)
> Compile divergence divergence tensor(4.9160e-07, device='cuda:0', dtype=torch.float64,
>        grad_fn=<MaxBackward1>)
> ```
> 
> As you can see, the results are closer, so we are more accurately computing this. Additionally, I'm not sure how valuable small divergences from gigantic inputs are. it's basically taking a negligible difference and compounding it. As documented, some amount of imprecision is to be expected: [pytorch.org/docs/stable/notes/numerical_accuracy.html#numerical-accuracy](https://pytorch.org/docs/stable/notes/numerical_accuracy.html#numerical-accuracy)
> 
> cc @mlazos - can you submit a public version of the fuzzer thing you wrote ? this would be great to reference and clear up theses sorts of issues - fallback_random as well.

Yeah this sounds good

**Comment by shaoyuyoung** [comment]
Hi, @eellison , thank you for the kind words and detailed comment. :)



> please compare the results to an fp64 reference input and only submit an issue if compile diverges more from fp64 than eager.

I'll pay attention next time. Just curiously, for the silent incorrectness, are there any criteria here to suggest that this is a bug? Because previouly I follow the instructions here:
https://github.com/pytorch/pytorch/issues/new?assignees=&labels=oncall%3A+pt2&projects=&template=pt2-bug-report.yml
![image](https://github.com/user-attachments/assets/d23a4ca1-3ab7-4b64-9087-a5ae511f3f27)
Perhaps we should update or develop a new standard to clarify these issues. It's also good for newcomers to pytorch and reduces some unnecessary issues. :)

Lastly, for this issue, `CPP BACKEND` seems to have some errors, maybe intel team can help take a look?

**Comment by eellison** [comment]
> If comparing eager and torch.compile at fp16/bf16, you should use fp32 as baseline
If the above requirements are met, add the label "topic: fuzzer" to your issue.

I will edit this so it says to compare to the higher precision. fp16/bf16 - use fp32. fp32 - use fp64. yes, I agree, a new standard would be helpful.

and sounds good - re-opening for cpu

**Comment by shaoyuyoung** [comment]
> fp_64_ref = c_model(x.double())
> print("Eager divergence", torch.max(torch.abs(output - fp_64_ref)))
> print("Compile divergence divergence", torch.max(torch.abs(c_output - fp_64_ref)))
> 
> Eager divergence tensor(0.0001, device='cuda:0', dtype=torch.float64, grad_fn=<MaxBackward1>)
> Compile divergence divergence tensor(4.9160e-07, device='cuda:0', dtype=torch.float64,
>        grad_fn=<MaxBackward1>)

I have done the same thing on CPU, and I got the following results:
```
Eager divergence tensor(7.5915e-07, dtype=torch.float64, grad_fn=<MaxBackward1>)
Compile divergence divergence tensor(2.2037e-05, dtype=torch.float64, grad_fn=<MaxBackward1>)
```
It seems to be the opposite of what happened on CUDA:  **The results are farther apart, so we are less accurately computing this.**

feel free to correct me if I am wrong.
